### PR TITLE
Fix crash caused by invalid `mix_rate` assignment due to bogus project settings.

### LINKED ISSUE #69819: `driver/mix_rate="Potato"` in project.godot crashes app
### Godot version

Godot beta 7

### System information

Ubuntu 22.04 - Nvidia GTX 970, Gnome shell 42 X11

### Issue description

```
[audio]
driver/mix_rate="Potato"
```
in project.godot cause crash in any project
```
drivers/pulseaudio/audio_driver_pulseaudio.cpp:226:135: runtime error: division by zero

================================================================
handle_crash: Program crashed with signal 8
Engine version: Godot Engine v4.0.beta.custom_build (45cac42c0b511672240c1fcccecfd3cc0580edcb)
Dumping the backtrace. Please include this when reporting the bug on: https://github.com/godotengine/godot/issues
[1] /mnt/Miecz/godot/bin/godot.linuxbsd.editor.dev.x86_64.san() [0x267209e] (/mnt/Miecz/godot/platform/linuxbsd/crash_handler_linuxbsd.cpp:56)
[2] /lib/x86_64-linux-gnu/libc.so.6(+0x3bcf0) [0x7f11f363bcf0] (??:0)
[3] AudioDriverPulseAudio::init_device() (/mnt/Miecz/godot/drivers/pulseaudio/audio_driver_pulseaudio.cpp:226 (discriminator 5))
[4] AudioDriverPulseAudio::init() (/mnt/Miecz/godot/drivers/pulseaudio/audio_driver_pulseaudio.cpp:350)
[5] AudioDriverManager::initialize(int) (/mnt/Miecz/godot/servers/audio_server.cpp:196)
[6] Main::setup2(unsigned long) (/mnt/Miecz/godot/main/main.cpp:2106 (discriminator 1))
[7] Main::setup(char const*, int, char**, bool) (/mnt/Miecz/godot/main/main.cpp:1882)
[8] /mnt/Miecz/godot/bin/godot.linuxbsd.editor.dev.x86_64.san(main+0x3b0) [0x2671886] (/mnt/Miecz/godot/platform/linuxbsd/godot_linuxbsd.cpp:61)
[9] /lib/x86_64-linux-gnu/libc.so.6(+0x23510) [0x7f11f3623510] (??:0)
[10] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x89) [0x7f11f36235c9] (??:0)
[11] /mnt/Miecz/godot/bin/godot.linuxbsd.editor.dev.x86_64.san(_start+0x25) [0x2671415] (??:?)
-- END OF BACKTRACE --
================================================================
Aborted (core dumped)

```

### Steps to reproduce

Above

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by Eoin-ONeill-Yokai**
I noticed this bug come in that was related to code I was already looking over for other reasons, so I decided to fix it.

We'll just default to a sensible value in the case that a user has somehow managed to modify the configuration file incorrectly. 44100hz feels like the appropriate default, but if there's a constant somewhere defined let me know and I'll change it to that. Otherwise, I'll change the code to use a constant so the fallback makes sense on a system-by-system basis and is easy to change.

It seems like, to me, there shouldn't be a case where we're loading the audio driver at all with an improper `mix_rate`, but I could be wrong. In which case, we will simply make exceptions for the `divide by zero` error. 

Closes #69819

### CODE REVIEW

File: drivers/pulseaudio/audio_driver_pulseaudio.cpp [hunk 1]
```diff
@@ -290,6 +290,11 @@ Error AudioDriverPulseAudio::init() {
 
 	mix_rate = GLOBAL_GET("audio/driver/mix_rate");
 
+	// In the case of invalid mix rate, let's default to a sensible value..
+	if (mix_rate <= 0) {
+		mix_rate = 44100;
```
Comment (adamscott) [inline drivers/pulseaudio/audio_driver_pulseaudio.cpp:295]: I suggest to add a warning that tells the user that their `audio/driver/mix_rate` setting is invalid and was set to 44 100 KHz.

File: drivers/xaudio2/audio_driver_xaudio2.cpp [hunk 2]
```diff
@@ -44,6 +44,12 @@ Error AudioDriverXAudio2::init() {
 	samples_in = nullptr;
 
 	mix_rate = GLOBAL_GET("audio/driver/mix_rate");
+
+	// In the case of invalid mix rate, let's default to a sensible value..
+	if (mix_rate <= 0) {
+		mix_rate = 44100;
```
Comment (adamscott) [inline drivers/xaudio2/audio_driver_xaudio2.cpp:50]: Same comment about the warning here.

**Comment by akien-mga** [comment]
I think there's a `DEFAULT_MIX_RATE` constant somewhere in AudioServer or AudioDriver which you could use.

I haven't checked but there are more audio driver implementations which might need the same treatment (alsa, coreaudio, wasapi, dummy).

**Comment by Eoin-ONeill-Yokai** [comment]
> I think there's a `DEFAULT_MIX_RATE` constant somewhere in AudioServer or AudioDriver which you could use.

Yes reusing that is a good plan. I've moved the declaration to a public section of the class.

> I haven't checked but there are more audio driver implementations which might need the same treatment (alsa, coreaudio, wasapi, dummy).

Yeah there was more to do there. I have added all the extra logic to the other platforms where appropriate. I skipped the warning in the dummy class, as it likely isn't as important there since users won't likely run into that specific driver all that often and it seems like it's for advance users anyway.

File: drivers/alsa/audio_driver_alsa.cpp [hunk 3]
```diff
@@ -44,7 +44,9 @@ extern int initialize_pulse(int verbose);
 #endif
 
 Error AudioDriverALSA::init_output_device() {
-	mix_rate = GLOBAL_GET("audio/driver/mix_rate");
+	StringName audio_driver_setting = "audio/driver/mix_rate";
+	mix_rate = GLOBAL_GET(audio_driver_setting);
```
Comment (akien-mga) [inline drivers/alsa/audio_driver_alsa.cpp:48]: Shouldn't this one get a validation check too?

**Review by akien-mga — COMMENTED** [review]
Since it's the same logic everywhere, it would make sense IMO to move this to a function like `AudioDriverManager::get_default_mix_rate()` so you don't need to duplicate the logic in each driver.

**Comment by Eoin-ONeill-Yokai** [comment]
> so you don't need to duplicate the logic in each driver.

I was wondering about this. Would it be ok to make a function that reads from the configuration and returns the default if it's invalid? It seems pretty specific and would only be useful in the constructor, but for a protected method perhaps it would be ok as long as people don't confuse it with `get_mix_rate` or similar functions. Like you said, it would cut down on repetition.

**Comment by akien-mga** [comment]
Yeah that's ok as a helper function. You can make it protected and even start with an underscore to make it clear that it's not part of the public API.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by YuriSizov** [comment]
Cherry-picked for 4.0.3.

**Pull request merged**